client/state/happychat/middleware.js

```diff
@@ -368,11 +289,11 @@ export default function( connection = null ) {
 				sendRouteSetEventMessage( connection, store, action );
 				break;
 
-			// NEW SOCKET API SURFACE
 			case HAPPYCHAT_IO_INIT:
-				connectionNG.init( store.dispatch, action.config );
+				connection.init( store.dispatch, action.auth );
```

Comment:
Do we really need to carry the auth handler through the action instead of just requiring it in the connection file?

Reply:
My rationale here is that we want `connection.init` to receive the data it uses, and be unaware of how the environment fetches that data (because different environments such as Calypso and dotorg will have different ways to get that). So, we want to pass this to the `connection.init`:

```
{
  signer_user_id,
  locale,
  groups,
  url,
  geoLocation,
  jwt
}
```

But because some of that data is async (geoLocation, jwt) I can only use a promise. 

Previously, we only called `connection.init` if the auth promise was fulfilled. What I didn't like about that is that it is more difficult to model the whole SocketIO flow across environments in a consistent way. In that scenario, the logic for the auth promise rejection is implemented by every environment and it's easy that it starts to diverge from one to another. With this approach, the whole initialization mechanism is modeled by the connection, and, and the same time, implementation details are left to the environment that uses the connection. Note also that making the connection unaware of the current auth mechanism is good for testing and for changing the auth mechanism depending on the environment as well.

I don't see a problem in passing a promise in the action, but I may be missing some context. Do you have some specific concerns in mind or does this rationale make sense?

Reply:
> My rationale here is that we want connection.init to receive the data it uses, and be unaware of how the environment fetches that data (because different environments such as Calypso and dotorg will have different ways to get that). So, we want to pass this to the connection.init:

I agree, connection should not contain env specific implementations.

> With this approach, the whole initialization mechanism is modeled by the connection, and, and the same time, implementation details are left to the environment that uses the connection.

My only concern is that "chained" promises can become very error prone. I guess by implementation details you mean each environment could catch its the auth errors ?

Because right now, if the sign request fails https://github.com/Automattic/wp-calypso/pull/19223/files#diff-81585a879ccd7f0f06a8723ffda5642bR23 it will through an unchaught in promise error. Arguably we could catch that error here https://github.com/Automattic/wp-calypso/pull/19223/files#diff-81585a879ccd7f0f06a8723ffda5642bR60 but that means that `getAuth()` would return a resolved promise ( on the catch side) which means it will go through the `then` branch in the connection implementation https://github.com/Automattic/wp-calypso/pull/19223/files#diff-9b65e4f7195e144f9194d6cf4e633800R49 which would throw errors when deconstructing the catched error.

Reply:
By implementation details, I meant how every environment implements the `auth` promise received by `connection.init`: Calypso, for example, would use the `get-happychat-auth.js`, but dotorg would not.

Good catch on how the auth promise is managed in get-happychat-auth! I should prepare a fix and add a test for this.